sbi/inference/snre/snre_base.py

```diff
@@ -352,39 +298,49 @@ def _get_loss(theta, x):
             log_prob_sum = 0
             with torch.no_grad():
                 for theta_batch, x_batch in val_loader:
-                    log_prob = _get_loss(theta_batch, x_batch)
+                    log_prob = self._loss(
+                        theta_batch, x_batch, clipped_batch_size, num_atoms
+                    )
                     log_prob_sum -= log_prob.sum().item()
                 self._val_log_prob = log_prob_sum / num_validation_examples
 
-            if self._show_progressbar:
-                # end="\r" deletes the print statement when a new one appears.
-                # https://stackoverflow.com/questions/3419984/
-                print("Training neural network. Epochs trained: ", epoch, end="\r")
-
-            if epoch >= max_num_epochs:
-                break
-
-        if self._show_progressbar and self._has_converged(epoch, stop_after_epochs):
-            # Network has converged, we print this summary.
-            print("Neural network successfully converged after", epoch, "epochs.")
-        elif self._show_progressbar and max_num_epochs == epoch:
-            # Training has stopped because of max_num_epochs argument.
-            print("Stopping neural network training after", epoch, "epochs.")
-
-        if max_num_epochs == epoch:
-            # warn if training was not stopped by early stopping
-            warnings.warn(
-                "Maximum number of epochs reached, but network has not yet "
-                "fully converged. Consider increasing the value of max_num_epochs."
-            )
+            self._maybe_show_progress(self._show_progressbar, epoch)
+
+        self._report_convergence_at_end(epoch, stop_after_epochs, max_num_epochs)
 
         # Update summary.
         self._summary["epochs"].append(epoch)
         self._summary["best_validation_log_probs"].append(self._best_val_log_prob)
 
-    @property
-    def summary(self):
-        return self._summary
+    def _classifier_logits(
+        self, theta: Tensor, x: Tensor, clipped_batch_size: int, num_atoms: int
+    ) -> Tensor:
```

Comment:
docstring, e.g.

Build atomic sets of (theta,x) pairs and return logits obtained through classifier forward pass.